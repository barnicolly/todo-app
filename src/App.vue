<template>
  <div class="todo-list">
    <h1 class="todo-list__header">Список дел</h1>
    <CAlert
      color="danger"
      :visible="alert !== ''"
      dismissible
      @close="closeAlert">
      <div>{{ alert }}</div>
    </CAlert>
    <AddForm class="todo-list__add-form"></AddForm>
    <ul>
      <li v-for="item in todoItems" :key="item.id">
        <ToDoItem
          :title="item.title"
          :completed="item.completed"
          :id="item.id"
          class="todo-list__item todo-item"></ToDoItem>
      </li>
    </ul>
    <small>{{ listSummary }}</small>
    <div>
      <small>
        Исходный код:&nbsp;
        <a target="_blank" href="https://github.com/barnicolly/vue-todo-app">
          https://github.com/barnicolly/vue-todo-app
        </a>
      </small>
    </div>
  </div>
</template>

<script setup lang="ts">
import ToDoItem from '@/components/ToDo/ToDoItem.vue';
import AddForm from '@/components/ToDo/AddForm.vue';
import { storeToRefs } from 'pinia';
import { useTodoStore } from '@/store/todo';
import { onMounted, computed } from 'vue';
import { CAlert } from '@coreui/vue';
import { useAlertStore } from '@/store/alert';

const store = useTodoStore();
const alertStore = useAlertStore();
const { todoItems } = storeToRefs(store);
const { alert } = storeToRefs(alertStore);

function closeAlert() {
  alertStore.removeAlert();
}

onMounted(async (): Promise<void> => {
  await store.fetchTodoItems();
});

const listSummary = computed(
  (): string =>
    `Выполнено задач: ${store.allCompletedToDoItems.length} из ${store.allToDoItems.length}`,
);
</script>
