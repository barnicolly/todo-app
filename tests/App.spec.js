import { flushPromises, mount } from '@vue/test-utils';

import axios from 'axios';
import App from '@/App.vue';
import { createTestingPinia } from '@pinia/testing';
import each from 'jest-each';

const mockTodoItems = [
  { id: '1', title: 'title1', completed: false },
  { id: '2', title: 'title2', completed: true },
];

each([
  ['all', mockTodoItems],
  ['active', mockTodoItems.filter((todo) => !todo.completed)],
  ['done', mockTodoItems.filter((todo) => todo.completed)],
]).test('filter %s', async (filter, todoItems) => {
  jest.spyOn(axios, 'get').mockResolvedValue({ data: todoItems });
  const wrapper = mount(App, {
    global: {
      plugins: [createTestingPinia({ stubActions: false })],
    },
  });
  expect(axios.get).toHaveBeenCalledTimes(1);
  await flushPromises();
  expect(wrapper.html()).toMatchSnapshot();
  jest.clearAllMocks();
});

describe('App.vue Snapshot', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('has same HTML structure', async () => {
    jest.spyOn(axios, 'get').mockResolvedValue({ data: mockTodoItems });
    const wrapper = mount(App, {
      global: {
        plugins: [createTestingPinia({ stubActions: false })],
      },
    });
    expect(axios.get).toHaveBeenCalledTimes(1);
    await flushPromises();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('has same HTML structure, empty list', async () => {
    jest.spyOn(axios, 'get').mockResolvedValue({ data: [] });
    const wrapper = mount(App, {
      global: {
        plugins: [createTestingPinia({ stubActions: false })],
      },
    });
    expect(axios.get).toHaveBeenCalledTimes(1);
    await flushPromises();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
