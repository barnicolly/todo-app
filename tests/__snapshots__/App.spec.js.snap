// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue Snapshot has same HTML structure 1`] = `
<div class="todo-list">
  <h1 class="todo-list__header">Список дел</h1>
  <transition-stub enterfromclass="" enteractiveclass="fade" entertoclass="fade show" leaveactiveclass="fade" appear="false" persisted="false" css="true">
    <!---->
  </transition-stub>
  <form class="todo-list__add-form"><input type="text" autocomplete="off" placeholder="Новая задача"><button class="btn btn--add" type="button"><svg class="svg-inline--fa fa-plus" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="plus" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512">
        <path class="" fill="currentColor" d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"></path>
      </svg></button></form>
  <ul>
    <li>
      <div class="todo-list__item todo-item"><input class="todo-item__checkbox" type="checkbox">
        <div class="todo-item__content"><label class="todo-item__label">Title1</label>
          <div class="todo-item__controls"><button class="btn btn--delete" type="button"><svg class="svg-inline--fa fa-trash-can" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="trash-can" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512">
                <path class="" fill="currentColor" d="M135.2 17.7C140.6 6.8 151.7 0 163.8 0H284.2c12.1 0 23.2 6.8 28.6 17.7L320 32h96c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 96 0 81.7 0 64S14.3 32 32 32h96l7.2-14.3zM32 128H416V448c0 35.3-28.7 64-64 64H96c-35.3 0-64-28.7-64-64V128zm96 64c-8.8 0-16 7.2-16 16V432c0 8.8 7.2 16 16 16s16-7.2 16-16V208c0-8.8-7.2-16-16-16zm96 0c-8.8 0-16 7.2-16 16V432c0 8.8 7.2 16 16 16s16-7.2 16-16V208c0-8.8-7.2-16-16-16zm96 0c-8.8 0-16 7.2-16 16V432c0 8.8 7.2 16 16 16s16-7.2 16-16V208c0-8.8-7.2-16-16-16z"></path>
              </svg></button></div>
        </div>
      </div>
    </li>
    <li>
      <div class="todo-list__item todo-item"><input class="todo-item__checkbox" type="checkbox">
        <div class="todo-item__content"><label class="todo-item__label">Title2</label>
          <div class="todo-item__controls"><button class="btn btn--delete" type="button"><svg class="svg-inline--fa fa-trash-can" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="trash-can" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512">
                <path class="" fill="currentColor" d="M135.2 17.7C140.6 6.8 151.7 0 163.8 0H284.2c12.1 0 23.2 6.8 28.6 17.7L320 32h96c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 96 0 81.7 0 64S14.3 32 32 32h96l7.2-14.3zM32 128H416V448c0 35.3-28.7 64-64 64H96c-35.3 0-64-28.7-64-64V128zm96 64c-8.8 0-16 7.2-16 16V432c0 8.8 7.2 16 16 16s16-7.2 16-16V208c0-8.8-7.2-16-16-16zm96 0c-8.8 0-16 7.2-16 16V432c0 8.8 7.2 16 16 16s16-7.2 16-16V208c0-8.8-7.2-16-16-16zm96 0c-8.8 0-16 7.2-16 16V432c0 8.8 7.2 16 16 16s16-7.2 16-16V208c0-8.8-7.2-16-16-16z"></path>
              </svg></button></div>
        </div>
      </div>
    </li>
  </ul><small>Выполнено задач: 1 из 2</small>
  <div><small> Исходный код:&nbsp; <a target="_blank" href="https://github.com/barnicolly/vue-todo-app"> https://github.com/barnicolly/vue-todo-app </a></small></div>
</div>
`;
